### ISSUE TITLE
Proposal: local coop handles should be pinned.

### ISSUE THREAD

**Issue by lambdageek** [mono/mono#15970]
# Summary

A reference to an object on the managed heap from a local (coop) handle will pin the object, allowing raw pointers `MonoObject*` to be passed around as long as the local handle is active.

# Motivation

Right now almost all coop handles (except those that are created in the icall marshalling layer using https://github.com/mono/mono/pull/11294/files#diff-51871b3c9f77c36adec88b19019cb69aR6395) are treated by the GC like normal managed reference and do not pin the object that they point to.

If handles are pinning, a lot of code will not need to be adapted to use coop handles - as long as the callee is only manipulating objects passed from the caller, it will not need to use handles.

The only functions that will need to use handles are places that pass managed references to callees: `foo (x->field)` where `field` is also a managed object.  The caller will need to create a local handle for `x->field` and before passing the raw pointer to `foo`:
```
   MonoObject *ptr = x->field;
   MonoObjectHandle h = MONO_HANDLE_NEW (MonoObject, ptr);
   foo (ptr);
```

# Proposed behavior

1. Change the behavior of the handles to pin the objects.
2. Allow an object for which a handle is active to be passed down to callers as a raw pointer to a `MonoObject` (or one of its subclasses).
3. Each caller may assume that that any arguments passed to it are pinned.
4. To return a managed object one of these strategies must be used:
  a. the caller must allocate a local handle and intialize it and then pass the handle (`MonoObjectHandleOut`) to the callee; or,
  b. the callee may be declared with a return type like `MonoObjectHandle` and use `HANDLE_FUNCTION_ENTER`/`HANDLE_FUNCTION_RETURN_REF`  to return a handle allocated in the caller's handle stack (ie same mechanism as today).

# Arguments against

## Missing handles are hard to pin down

With this proposal, it will be difficult to find places where we forgot to pin some object. 
Right now any code that manipulates `MonoObject*` directly is suspect.  We're in the process of rewriting all of it to use the macros from `handles.h`.  If we adopt this proposal, we will have a lot of `x->y->z` raw object manipulation that is benign.  It will be more difficult to spot the needle in the haystack.  This is similar to the issues we've had in locating missing write barriers: `x->f = y` looks innocent and it's easy to forget to write `mono_object_setref (x, f, y)` instead.

# Details

## Fully preemptive suspend

(Fully preemptive suspend is Mono's historical behavior - it's what we had with Boehm and it's what we used by default with Sgen initially.)

With preemptive suspend we're committed to conservative stack scanning no matter what - we could interrupt code while the only reference to a live object is in a register.  We don't know anything, so anything can be a pointer.

This proposal won't change anything about preemptive suspend.

## Hybrid suspend and coop suspend

(In hybrid suspend, Mono cooperatively suspends managed code and native runtime code at well defined safepoints.  Code that we don't control (which could nonetheless call back into Mono) is preemptively suspended.  All embedding API functions perform thread state transitions)

(In full coop suspend, Mono only suspends at safepoints - external code that we don't control continues running while the GC is doing a collection. So any external code either has to not interact with the runtime at all (like syscalls), or it must use thread state transitions API to switch between GC Safe and GC Unsafe mode explicitly thereby inserting safepoints. Full coop is used only in scenarios where Mono is alone or with an embedder that is closely aware of the runtime.)

Currently we code pretty defensively in the native runtime code: we assume that any callee could be hiding a GC safepoint, and consequently that any raw pointers that live across a function call may become invalidated because the GC moved the objects.

With this proposal, coop handles give us a guarantee that as long as a handle exists, the raw pointer is valid.  As a result a lot of the handle macros could simplified or unnecessary.

**Comment by lambdageek** [comment]
## Precise GC

I initially thought this proposal was going to make it more difficult to stop conservatively scanning the native stack.  But I think actually it's ok.  

To do precise GC we need two things:
1. descriptors for the managed stack frames
2. a guarantee that any object touched by unmanaged code has a reference that's visible to the GC.

Without this proposal we also have:
3. no unmanaged code holds a raw pointer to a managed object across a safepoint.

But actually we don't care about (3) (except in so much as we care about nursery fragmentation) - what we really want is a modification of (2):
2a. if unmanaged code has a reference to a raw object, there's at least one pinning reference to that object that's visible to the GC.

**Comment by lambdageek** [comment]
Thanks @jaykrell and @vargaz for championing this.

**Comment by lambdageek** [comment]
The CoreCLR way of doing this is described in https://github.com/dotnet/coreclr/blob/a28b25aacdcd2adb0fdfa70bd869f53ba6565976/Documentation/botr/mscorlib.md#calling-from-managed-to-native-code

**Comment by jaykrell** [comment]
If I have a handle to x, do I need a handle to x->field?

Is C++ still being considered? It is maybe related.

Is this proposal considered temporary or permanent?

If accepted, should coop conversion continue with the same “complete” goal?

Pros: handles are not free. This allocates fewer.

If accepted, will typedefs or comments be used to indicate human review occurred and code is believed safe?
Seems cheap & easy but not ironclad.

And static analysis updated?

**Comment by jaykrell** [comment]
I wonder also if the JIT icalls (`MONO_HANDLE_REGISTER_ICALL` at the bottom of `icall-def.h`) should get more similar treatment as the other icalls (`HANDLES` in `icall-def.h`). i.e. handle is pointer to volatile local and no TLS access.

**Comment by vargaz** [comment]
I agree with this proposal. There are drawbacks, but i think the advantages outweigh them.

**Comment by jaykrell** [comment]
The code currently combines precisely finding roots, with pinning?
This new case is we'd precisely find them, but not every instance of every root, as I understand.

**Comment by lambdageek** [comment]
> If I have a handle to x, do I need a handle to x->field? (Make x volatile?)

Suppose `x` is a `MonoReflectionModule*`
1. If you have a handle to `x`, `x` is pinned.
2. While you have a handle to `x`, you can use `&x->token` which is a `guint32*` however you want. So `x->token = 42` and `printf ("%x", x->token)` are both okay.
3. While you have a handle to `x` you can use `&x->image` which is a `MonoImage**` however you want.
4. While you have a handle to `x` you may `MONO_OBJECT_SETREF_internal (x, name, newName)` (`x->name` is a `MonoString*`) where `newName` is some other `MonoString*`.
5. While you have a handle to `x` in order to read `x->name` you have to make a handle to it.

I don't think we need to make `x` volatile (what am I missing?).

I also don't think that we should be making  `MonoReflectionModule:name` as `MonoString * volatile name;` because the first access to `x->name` should be to create a handle:

```
   void
   my_func (MonoReflectionModule *x)
   {
      /* callee assumes caller pinned x with a handle */
     HANDLE_FUNCTION_ENTER ();
     MonoStringHandle x_name = MONO_HANDLE_NEW (MonoString, x->name);
     /* Now can use x->name or MONO_HANDLE_RAW (x_name) interchangeably, assuming we don't expect another user thread to modify the name */
     ...
     HANDLE_FUNCTION_RETURN ();
   }
```

**Comment by jaykrell** [comment]
Makes sense. I suggest all managed fields be renamed, so only blessed access can see them. But it is a heavy hammer.

**Comment by jaykrell** [comment]
Thoughts on pinned vs. non-pinning coop handles, long term?

i.e.

 - finish short term pinned work, then resume old non-pinned work?

 - finish short term pinned work, then undo old non-pinned work?

 - finish short term pinned work, then leave alone old non-pinned work?

 - how to write new code?

**Comment by jaykrell** [comment]
Also I suggest a little more crispness regarding "root discovery" and "pinning".
It is still precise. All roots are found. Without just blindly reading all of of the stack. But not all the instances of the roots are found.
